From 5760adcc278aea791b325aa0b63a44af4c9ac990 Mon Sep 17 00:00:00 2001
From: Gregory Haskins <ghaskins@novell.com>
Date: Thu, 22 Oct 2009 14:22:53 -0400
Subject: [PATCH 073/119] venetdev: optimize rx-empty wakeup

This WQ is almost always going to be empty, yet we try to do a wakeup
on it any time the rx thread is about to go idle.  Lets optimize this
to avoid the wqh.spinlock.

Signed-off-by: Gregory Haskins <ghaskins@novell.com>
---
 kernel/vbus/devices/venet/device.c |    3 ++-
 1 files changed, 2 insertions(+), 1 deletions(-)

diff --git a/kernel/vbus/devices/venet/device.c b/kernel/vbus/devices/venet/device.c
index 13c1611..058830d 100644
--- a/kernel/vbus/devices/venet/device.c
+++ b/kernel/vbus/devices/venet/device.c
@@ -1024,7 +1024,8 @@ next:
 	if (ioq_empty(ioq, ioq_idxtype_inuse)) {
 		clear_bit(RX_SCHED, &priv->flags);
 		ioq_notify_enable(ioq, 0);
-		wake_up(&priv->vbus.rx_empty);
+		if (waitqueue_active(&priv->vbus.rx_empty))
+			wake_up(&priv->vbus.rx_empty);
 	}
 
 	vbus_connection_put(conn);
-- 
1.6.5.2

